Fix panic in loki source file

wildum opened the pull request:
<!--

CONTRIBUTORS GUIDE: https://github.com/grafana/alloy/blob/main/docs/developer/contributing.md#updating-the-changelog

If this is your first PR or you have not contributed in a while, we recommend
taking the time to review the guide. It gives helpful instructions for
contributors around things like how to update the changelog.

-->

#### PR Description

The panic would happen when the runner would need to stop before it has time to start running

#### Which issue(s) this PR fixes

<!-- Uncomment the following line if you want that GitHub issue gets automatically closed after merging the PR -->
<!-- Fixes #issue_id -->

#### Notes to the Reviewer

#### PR Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] CHANGELOG.md updated
- [x] Tests updated

ptodev commented on internal/component/loki/source/file/tailer.go:
```diff
@@ -348,14 +348,14 @@ func (t *tailer) Stop() {
+	var err error
 	// Stop the underlying tailer to prevent resource leak.
 	if t.tail != nil {
+		// Save the current position before shutting down tailer
```
Would be nice to have a comment on why this snippet is here.

wildum replied:
I updated the comment

wildum commented:
> do any of these changes need to be ported over to the decompresser for compressed files?

Good catch, I took a similar approach for the decompresser and added a test.

> have we tested with file rotation?

I ran locally on Windows and Mac stress scenarios where thousands of files are created, deleted, rotated, and written at a very high frequency for a few minutes. Then I made sure that Alloy would still correctly tail all files with a script.

aglees commented:
I think there is still an issue with this. We've had to rollback `v1.8.1` to `v1.6.1` in our churning DEV environment because logs on short-lived nodes are being missed. We're seeing something like the initial Daemonset pods getting logs captured, but it appears that we're still getting a panic and no further container logs being recorded.

dehaansa commented:
@aglees I'm sorry to hear that you're still encountering issues. Can you open a new issue with information about your environment, your configuration, and the panic that you are seeing?
